[DOC-3661][DOC-3663] Add a concept for CMEK (#13963)
cockroachdb/docs: cockroachcloud/managing-cmek.md

docs_area: manage.security
---

> Customer-Managed Encryption Keys (CMEK) for {{ site.data.products.dedicated }} allows the customer to delegate responsibility for the work of encrypting their application data to Cockroach Labs, while maintaining the ability to completely revoke Cockroach Labs' access with a single operation.

Reviewer: ```suggestion
Customer-Managed Encryption Keys (CMEK) for {{ site.data.products.dedicated }} allows the customer to delegate responsibility for the work of encrypting their cluster data to CockroachDB Dedicated, while maintaining the ability to completely revoke CockroachDB Dedicated's access with a single operation.
```

{{site.data.alerts.callout_info}}
This feature is a Private Preview release, limited to customers for whom it is enabled, upon their request.
{{site.data.alerts.end}}

This page guides the user through the process of enabling Customer-Managed Encryption Keys (CMEK) for your {{ site.data.products.dedicated }} cluster.

See also:

- [Customer-Managed Encryption Key (CMEK) frequently asked questions (FAQ)](cmek-faq.html)
- [Encryption at Rest (enterprise)](../{{site.versions["stable"]}}/security-reference/encryption.html#encryption-at-rest-enterprise)

## Overview of CMEK management procedures

This section gives a high level sketch of the operations involved with implementing CMEK with your CockroachDB cluster.

- [Enabling CMEK](#enable-cmek) for a {{ site.data.products.dedicated }} requires several steps:

	- Turning on the {{ site.data.products.dedicated }} feature.

> 	- Provisioning an encryption key with your infrastructure as a service (IaaS) provider's key management service (KMS).

Reviewer: ```suggestion
	- Provisioning an encryption key with your cloud provider's key management service (KMS).
```

	- Granting {{ site.data.products.db }} access to the CMEK i.e. the key you control.
	- Switching your cluster to use the CMEK.

- [Checking CMEK status](#check-cmek-status) allows you to inspect the CMEK state of your cluster with a call to the Cluster API.

- [Revoking CMEK for a cluster](#revoking-cmek-for-a-cluster) by revoking Cockroach Labs' access to your CMEK at the IAM/KMS level.

- [Restore CMEK following a revocation event](#restore-cmek-following-a-revocation-event) by reauthorizing {{ site.data.products.db }} to use your key, and coordinating with our service team to assist in recovering your Organization.

## Enable CMEK

### Step 1. Prepare your {{ site.data.products.dedicated }} Organization

1. To start the process, contact Cockroach Labs by reaching out to your account team, or [creating a support ticket](https://support.cockroachlabs.com/). You must provide your **Organization ID**, which you can find in your [Organization settings page](https://cockroachlabs.cloud/settings).

	They will enable the [Cloud API](cloud-api.html) and CMEK for your {{ site.data.products.dedicated }} Organization.

	{{site.data.alerts.callout_info}}

> 	The upgrade process will generate a new Organization ID. You will need this new Organization ID when provisioning the crucial cross-account IAM role that will allow Cockroach Labs to encrypt/decrypt on your behalf with the CMEK.

Reviewer: Not needed anymore as it was before cloud API was in preview.

	{{site.data.alerts.end}}

1. [Create a {{ site.data.products.db }} service account](console-access-management.html#service-accounts).

1. [Create an API key](console-access-management.html#create-api-keys) for the service account to use.

### Step 2. Provision your cluster

1. Create a new {{ site.data.products.dedicated }} cluster. You can do this either of two ways:
	- [Using the {{ site.data.products.db }} console clusters page](https://cockroachlabs.cloud/cluster).
	- [Using the Cloud API](cloud-api.html#create-a-new-cluster). 

### Step 3. Provision IAM and KMS in your IaaS

Next, you must provision the resources required resources in your IaaS, whether this is AWS or GCP:

1. The cross-account IAM role that Cockroach Labs will use to access your key.

1. The key itself.

> Follow the instructions depending on your IaaS:

Reviewer: ```suggestion
Follow the instructions depending on your cloud provider:
```

## Revoking CMEK for a cluster

> Revoking access to the CMEK means disabling all encryption/decryption of data in your database, which means preventing reading and writing any data from or to your database. This likely implies a shutdown of your service, with significant business implications. This is a disaster mitigation tactic to be used only in scenario involving a severe, business critical security breach.

Reviewer: ```suggestion
Revoking access to the CMEK means disabling all encryption/decryption of data in your cluster, which means preventing reading and writing any data from or to your cluster. This likely implies a shutdown of your service, with significant business implications. This is a disaster mitigation tactic to be used only in scenario involving a severe, business critical security breach.
```

> This can be done temporarily or permanently. This action is performed at the level of your infrastructure (IaaS).

Reviewer: ```suggestion
This can be done temporarily or permanently. This action is performed at the level of your cloud provider.
```

### Step 1: Revoke IAM access

{{site.data.alerts.callout_danger}}
Do not delete the CMEK key.
Deleting the CMEK key will permanently prevent decryption of your data, preventing all possible access and rendering the data inaccessible.
{{site.data.alerts.end}}

> First, revoke Cockroach Labs' access to your key at the IAM level with your infrastructure provider. 

Reviewer: ```suggestion
First, revoke CockroachDB Dedicated's access to your key at the IAM level with your cloud provider. 
```

This will **not** immediately stop your cluster from encrypting and decrypting data, which does not take effect until you update your cluster in the next step.

That is because CockroachDB does not use your CMEK key to encrypt/decrypt data, but only to encrypt/decrypt a key encryption key (KEK). The KEK is used to encrypt a data encryption key (DEK), which is used to encrypt/decrypt your application data.

Your cluster will continue to use the already-provisioned DEK until you make the Cloud API call to revoke CMEK.

> ### Step 2: Update your cluster to stop using the CMEK key for encryption

Reviewer: ```suggestion
### Step 2: Update your cluster to stop using the CMEK key
```

> Your cluster will continue to operate with the encryption keys it has provisioned with your CMEK key until you order it to switch off of CMEK, and use a Cockroach Labs managed encryption key instead.

Reviewer: ```suggestion
Your cluster will continue to operate with the encryption keys it has provisioned with your CMEK key until you instruct it to switch off CMEK.
```
